solarforecastarbiter/metrics/event.py

```diff
@@ -0,0 +1,230 @@
+    if isinstance(fx, (list, tuple)):
+        fx = np.array(fx)
+
+    if len(obs) == 0:
+        raise RuntimeError("No Observation timeseries data.")
+    elif len(fx) == 0:
+        raise RuntimeError("No Forecast timeseries data.")
+
+    tp = np.count_nonzero(np.logical_and(fx, obs))
+    fp = np.count_nonzero(np.logical_and(fx, ~obs))
+    tn = np.count_nonzero(np.logical_and(~fx, ~obs))
+    fn = np.count_nonzero(np.logical_and(~fx, obs))
+    return tp, fp, tn, fn
+
+
+def probability_of_detection(obs, fx):
+    """Probability of Detection (POD).
+
+    .. math:: \\text{POD} = \\text{TP} / (\\text{TP} + \\text{FN})
+
+    Parameters
+    ----------
+    obs : (n,) array-like
+        Observed event values (True=event, False=no event).
+    fx : (n,) array-like
+        Forecasted event values (True=event, False=no event).
+
+    Returns
+    -------
+    pod : float
+        The POD of the forecast.
+
+    """
+
+    tp, fp, tn, fn = _event2count(obs, fx)
+    if (tp + fn) == 0:
+        return 0.0
+    else:
+        return tp / (tp + fn)
+
+
+def false_alarm_ratio(obs, fx):
+    """False Alarm Ratio (FAR).
+
+    .. math:: \\text{FAR} = \\text{FP} / (\\text{TP} + \\text{FP})
+
+    Parameters
+    ----------
+    obs : (n,) array-like
+        Observed event values (True=event, False=no event).
+    fx : (n,) array-like
+        Forecasted event values (True=event, False=no event).
+
+    Returns
+    -------
+    far : float
+        The FAR of the forecast.
+
+    """
+    tp, fp, tn, fn = _event2count(obs, fx)
+    if (tp + fp) == 0:
+        return 0.0
+    else:
+        return fp / (tp + fp)
+
+
+def probability_of_false_detection(obs, fx):
+    """Probability of False Detection (POFD).
+
+    .. math:: \\text{POFD} = \\text{FP} / (\\text{FP} + \\text{TN})
+
+    Parameters
+    ----------
+    obs : (n,) array-like
+        Observed event values (True=event, False=no event).
+    fx : (n,) array-like
+        Forecasted event values (True=event, False=no event).
+
+    Returns
+    -------
+    pofd : float
+        The POFD of the forecast.
+
+    """
+
+    tp, fp, tn, fn = _event2count(obs, fx)
+    if (fp + tn) == 0:
+        return 0.0
+    else:
+        return fp / (fp + tn)
+
+
+def critical_success_index(obs, fx):
+    """Critical Success Index (CSI).
+
+    .. math:: \\text{CSI} = \\text{TP} / (\\text{TP} + \\text{FN} + \\text{FN})
+
+    Parameters
+    ----------
+    obs : (n,) array-like
+        Observed event values (True=event, False=no event).
+    fx : (n,) array-like
+        Forecasted event values (True=event, False=no event).
+
+    Returns
+    -------
+    csi : float
+        The CSI of the forecast.
+
+    """
+
+    tp, fp, tn, fn = _event2count(obs, fx)
+    if (tp + fp + fn) == 0:
+        return 0.0
+    else:
+        return tp / (tp + fp + fn)
+
+
+def event_bias(obs, fx):
+    """Event Bias (EBIAS).
+
+    .. math:: \\text{EBIAS} = (\\text{TP} + \\text{FP}) / (\\text{TP} + \\text{FN})  # NOQA
```

Comment:
what's the reference for this? I'm struggling to find it.

Reply:
Let me check on this.

Reply:
I reviewed the copies of the proposed metrics document that I have/can find and there is no included reference for the event bias (or most of the event metrics). I did find some pages via Google that describe bias defined for categorical events, but no clearly "good" reference. I'll talk with Qin Wang (EPRI) since he wrote the origin draft of the metrics document and then follow-up after.

Reply:
Update: Qin provided two references that he used for the event forecast metrics portion:
- Jensen, T. L., Fowler, T. L., Brown, B. G., Lazo, J. K., & Haupt, S. E. (2016) "Metrics for Evaluation of Solar Energy Forecasts", (No. NCAR/TN-527+STR). doi: 10.5065/D6RX99GG
- Chapter 8 "Forecast Verification" of D. S. Wilks "Statistical Methods in the Atmospheric Sciences"

Although I haven't confirmed whether "EBIAS" is standard nomenclature, the given definition is discussed on page 310 in the "Forecast Verification" chapter of Wilks (which just calls it "bias") and in page 13/table 9 of Jensen et al. (2016), where it's called "frequency bias (FBIAS)".